DRAFT: Guidance on CHERI-enabled allocator API guarantees and implementations (#51)
CTSRD-CHERI/cheri-c-programming: src/allocators/allocator-recommendations.md

## Recommendation for allocator implementations

This section provides recommendations for the implementers of heap
allocators on CHERI-enabled systems.
There is also specific guidance on handling alignment and padding required to
take into account bounds imprecision in [Implications for memory-allocator
design](../apis/implications-for-memory-allocator-design.md).

### Allocating memory

In addition to implementing the conventational invariants ensuring the
mutually exclusive allocation of memory, CHERI-aware implementations of
`malloc()`, `calloc()`, and `posix_memalign()` must either return a capability
holding a `NULL` pointer (on failure), or a capability holding a non-`NULL`
pointer (on success) that:

| Requirement | Rationale |
|-------------|-----------|
| Is valid (i.e., with its tag bit set) | Pointer must be dereferenceable for memory accesses on return |
| Is unsealed | Pointer must be dereferenceable for memory access on return |
| Has bounds that permit access to the full requested range of the allocation | Full allocation must be accessible for memory access on return |

> | Has bounds that do not permit access to any other current allocation, nor to allocator metadata, implementing non-aliasing spatial safety | Allow access only to the memory allocation itself, and not either allocator-internal state (enabling a long history of heap corruption attacks or the leakage of privileged capabilities) or other allocations (violating non-aliasing requirements for spatial safety) |

Reviewer: ```suggestion
| Has bounds that do not permit access to anything other than the allocated memory | Allow access only to the memory allocation itself, and not either allocator-internal state (enabling a long history of heap corruption attacks or the leakage of privileged capabilities) or other allocations (violating non-aliasing requirements for spatial safety) |
```
Shorten requirement as is duplicated in the rationale

> | Has permissions that allow data load, data store, capability load, and capability store | C/C++ expect that memory will not only be accessible for read and write, but also that pointers can be held in arbitrary heap memory.  While it is possible (and reasonable) to imagine heap allocators that return pointers without the ability to load or store capabilities, doing so with the general-purpose C/C++ allocators will break almost all extant software. |

Reviewer: We might want to relax this in a type-aware allocator world.

> | Be sufficiently aligned to allow capability loads and stores at relative offset 0 from the returned pointer | As with capability load and store positions, being unable to use pointers stored at pointer alignment within an alllocation will break almost all extant software. |

Reviewer: As an incidental feature of revocation we don't allow < 16-byte allocations. If we did, this requirement wouldn't hold. I'm not sure if we want to point that out.

Reviewer 2: I don't think we need to require this for allocations < cap_size. You could have an allocator that has a bucket just for integers allocations. Maybe qualify the requirement with "for allocations >= cap_size"?

Also I believe there is a typo here? positions -> permissions?

Author: @arichardson: Typo corrected, thanks; will respond to the remainder of this thread separately.

Author: I've added a note about capability alignment being required only for allocations of at least pointer size.

Author: Probably this should in fact be capability size.

Author: I wonder if there's an interesting implication for `realloc()` here: Should `realloc()` always ensure at least pointer alignment if reallocation is requested to go from below to at or above capability size, meaning that it might need to allocate / copy memory? I think this may be implied currently, but perhaps should be stated explicitly... if we think it's useful / important.

Reviewer: You could restate the requirement on malloc that all allocations that could hold a capability be capability aligned. It's required that memory allocated by realloc be interchangeable with memory allocated by malloc so it's IMO a pretty obvious requirement.

The allocator must:

| Requirement | Rationale |
|-------------|-----------|

> | Pad below and above allocations such that, when precise bounds are not utilized, no other allocation is accessible within returned bounds | This is a corallary of the above requirement to set bounds to prevent access to allocator metadata and other allocations, non-aliasing in the presence of bounds imprecision requires padding so that accessible memory outside of the requested size cannot access disallowed data. |

Reviewer: ```suggestion
| Pad below and above allocations such that, when the requested allocation does not permit precise bounds, no other allocation is accessible within returned bounds | This is a corollary of the above requirement to set bounds to prevent access to allocator metadata and other allocations, non-aliasing in the presence of bounds imprecision requires padding so that accessible memory outside of the requested size cannot access disallowed data. |
```
I keep reading the previous version as the allocator implementor using inexact bounds which isn't the case. Instead, the implementor pads sufficiently to use exact bounds.

Reviewer 2: ```suggestion
| Ensure no other allocation is accessible within returned bounds by padding below and above allocations if the requested allocation size does not permit precise bounds | This is a corollary of the above requirement to set bounds to prevent access to allocator metadata and other allocations, non-aliasing in the presence of bounds imprecision requires padding so that accessible memory outside of the requested size cannot access disallowed data. |
```
I find it a bit easier to read with the order of this sentence swapped?

The allocator should:

| Requirement | Rationale |
|-------------|-----------|
| Fill reachable memory within bounds with zeroes before returning a pointer to it | Prevent both uninitialized values and data remanence by deterministically overwriting data that may otherwise be residual from previous allocations or memory-allocator metadata. |

The allocator must select one of the following two choices for each memory allocation; different memory allocations may use different strategies (e.g., enabling precise bounds for smaller allocations, or imprecise ones for larger allocations, as well as other strategies):

| Requirement | Rationale |
|-------------|-----------|

> | Provide precise bounds -- i.e., in which the lower bound is the lowest address of the returned allocation, and the upper bound is the highest address of the returned allocation plus one | Implement precise bounds, where achievable whtin the constraints of bounds imprecision, avoiding the use of padding while still allowing access to the full allocation. |

Reviewer: ```suggestion
| Provide precise bounds -- i.e., in which the lower bound is the lowest address of the returned allocation, and the upper bound is the highest address of the returned allocation plus one | Implement precise bounds, where achievable within the constraints of bounds imprecision, avoiding the use of padding while still allowing access to the full allocation. |
```

> | Provide precise bounds -- i.e., in which the lower bound is the lowest address of the returned allocation, and the upper bound is the highest address of the returned allocation plus one | Implement precise bounds, where achievable whtin the constraints of bounds imprecision, avoiding the use of padding while still allowing access to the full allocation. |

Reviewer 2: This seems like it is just repeating the section above? Maybe better have this as a separate "ensuring non-overlapping bounds" section rather than two tables?

> | Pad and/or align allocation such that the returned address is equal to the lower bound | Implement non-aliasing memory by passing allocations and setting bounds to include only the allocation and its padding. |

Reviewer: Given this is a discussion for malloc(), etc, this section doesn't make sense to me. You can have a programming environment that requires the requestor make things precisely representable, but that's not something we've implemented and would be extremely incompatible.

Author: Hmm. I've probably expressed this poorly -- the intention is to say that when setting up an individual allocation, one of these two strategies must be selected, not that the same strategy should be used for every allocation. I.e., you can either be precise, or you can be imprecise and use padding.

Reviewer: I think I may be caught by reading your use of "select" as though it's a choice. I find that framing confusing. In practice the allocator computes how much padding is required for exact representation (which might be zero) and then allocates at least that much space (possibly adding padding due to allocator design choices).

The allocator may:

| Requirement | Rationale |
|-------------|-----------|
| Implement fail-stop semantics if the call fails for one or more of the above reasons | Fail stop rather than potentially allow internal allocator state to be corrupted; this will also prevent entering potentially harmful undefined behaviour in the caller. |
| Disregard invalid `free()` requests, continuing execution | Ignore the improper request rather than potentially allow internal allocator state to be corrupted, enabling improved availability at the cost of potential memory leakage and entry of undefined states in the caller. |
| Implement diagnostic features such as logging, crash dumps, and so on when such usage a failure occurs | This design choice applies given either of the above choices, and is recommended to improve diagnostic capacity in the system. |

Relating specifically to detecting and preventing use-after-free and
use-after-reallocation, the allocator may:

| Requirement | Rationale |
|-------------|-----------|

> | Implement pointer revocation by clearing the tag of capabilities pointing to freed memory | Clearing the tag ensures non-dereferenceability when a pointer is revoked, althugh this can also be achieved by other means such as changing page-table permissions on the page a pointer points to, on supporting architectures. |

Reviewer: ```suggestion
| Implement pointer revocation by clearing the tag of capabilities pointing to freed memory | Clearing the tag ensures non-dereferenceability when a pointer is revoked, although this can also be achieved by other means such as changing page-table permissions on the page a pointer points to, on supporting architectures. |
```
I feel like the page table option doesn't belong in the rational of tag clearing and should be it's own requirement.

> | On virtual-memory-enabled systems, unmap reachable memory within the bounds of the allocation after it has been freed | This may be useful both for rapidly detecting use-after-free errors and also to allow the allocator to reuse physical memory while deferring revocation of capabilities to specific virtual addresses. |

Reviewer: I wonder if we want to include page permissions here. I worry that unmapped doesn't quite describe the state we want (backing store freed with fault on access). I think one could achieve the same effect with MADV_FREE and mprotect(PROT_NONE).

| Revoke capabilities to the storage immediately upon free | This is not required as larger CHERI systems will not be able to implement this behaviour efficiently, but it is permitted as this will tighten implementations from non-aliasing emporal safety to precise temporal safety, which may be efficient in smaller designs. |

### Reallocating memory

The allocator must ignore, or trap on, calls to `realloc()` if the passed
capability:

| Requirement | Rationale |
|-------------|-----------|
| Violates any of the properties checked for in `free()` | In effect, reallocation is a request to free memory, and will typically be implemented by the same paths within the allocator. |

The allocator must:

| Requirement | Rationale |
|-------------|-----------|
| Return a capability with the same properties as those defined for `malloc()` | In effect, reallocation is a request to allocate new heap memory, and will typically be implemented by the same paths within the allocator. |

The allocator must not:

| Requirement | Rationale |
|-------------|-----------|

> | Return a new pointer from `realloc()` that has an identical address to the passed argument but differs in its bounds or other metadata | This requirement arises from the observation that callers to `realloc()` will immediately compare the passed pointer to the returned one to establish whether the memory was reallocated or not.  If it was, it is frequently the case that they will use the passed pointer rather than the returned one, since they are assumed to be interchangeable.  With CHERI C/C++, pointer equality testing with the `==` operator compares only the address and not bounds or other metadata, and as a result a pointer with the prior bounds, rather than the new bounds, may frequently be used.  Rather than allowing this situation to occur, we require reallocation.  This tradeoff point appears to maximise compatibility at the cost of performance and memory overhead.  It may be that this requirement is weakened to "should not" in the future. |

Reviewer: We should probably make it explicit that this is catering to de facto C/C++ and that such program are relying on UB.

Author: Please feel free to suggest wording.

 * Has bounds that do not permit access to any other current allocation,
   implementing non-aliasing spatial safety
 * Has permissions that allow data load, data store, capability load, and
   capability store
 * Be sufficiently aligned to allow capability loads and stores at relative
   offset 0 from the returned pointer

The allocator should:

 * Zero the allocation before returning a pointer to it

The allocator may:

 * Provide precise bounds -- i.e., in which the lower bound is the lowest
   address of the returned allocation, and the upper bound is the highest
   address of the returned allocation plus one

### Freeing memory

CHERI-aware allocators must ignore, or trap on, calls to `free()` if the
passed capability:

 * Is invalid (i.e, with its tag bit unset); or
 * Is sealed; or
 * Has bounds that have been changed from those returned by the allocator
   when the allocation was first made; or
 * Has permissions that disallow any of data load, data store, capability
   load, or capability store

The allocator may implement fail-stop semantics if the call fails for one or
more of the above reasons.

> 

Reviewer 3: It is probably safe to leave up to the allocator whether a nonzero offset between address and base is acceptable.

The CHERIoT allocator overloads `free()` (well, `heap_free()`) to permit dropping claims on an object, and in that case it accepts interior, subset, &| attenuated pointers.  That is, claims may be (taken and) dropped on subobjects (or, really, arbitrary slices of objects) but deallocation requires the original pointer (a la C).

Author: Regarding the former: Are there use cases for non-zero offsets in heap allocators we are aware of beyond, perhaps, "deterministically trap on overflow rather than on underflow when imprecise bounds require padding"?

Author: I've now added text to this effect.

This section provides recommendations for the implementers of heap
allocators on CHERI-enabled systems.
There is also specific guidance on handling alignment and padding required to
take into account bounds imprecision in [Implications for memory-allocator
design](../apis/implications-for-memory-allocator-design.md).

### Allocating memory

In addition to implementing the conventational invariants ensuring the
mutually exclusive allocation of memory, CHERI-aware implementations of
`malloc()`, `calloc()`, and `posix_memalign()` must either return a capability
holding a `NULL` pointer (on failure), or a capability holding a non-`NULL`
poiner (on success) that is:

 * Valid (i.e., with its tag bit set)
 * Unsealed
 * Has bounds that permit access to the full requested range of the allocation
 * Has bounds that do not permit access to any other current allocation, nor
   to allocator metadata, implementing non-aliasing spatial safety
 * Has permissions that allow data load, data store, capability load, and
   capability store
 * Be sufficiently aligned to allow capability loads and stores at relative
   offset 0 from the returned pointer

The allocator should:

>  * Fill reachable memory within bounds with zeroes before returning a pointer

Reviewer 2: I feel like this is really a non-CHERI-specific recommendation? I guess maybe more relevant to ensure you don't have stale capabilities?

Author: I wonder if we need to spell out more specifically that we are recommending allocator behaviours to also address uninitialised data aspects of temporal safety.

Author: I have added notes on this in new rationale text.

Author: @nwf, @qwattash: Do we want to draw out a specific benefit relating to capabilities flowing through uninitialised data from the perspective of use-after-free, etc? I think right now that's not a key part of the model as we revoke memory in quarantine not just out of quarantine, but that doesn't make it a generally good thing to do.

   to it

The allocator may:

>  * Provide precise bounds -- i.e., in which the lower bound is the lowest

Reviewer 2: Isn't this more of a "The allocator does not necessarily"? Since the properties here are basically what is described by the first enumeration?

Author: The aim of this and the above text were to require non-aliasing spatial safety but to permit precise memory safety. Maybe that framing is confusing rather than helpful.

Author: I have added notes on this in new rationale text.

   address of the returned allocation, and the upper bound is the highest
   address of the returned allocation plus one

### Freeing memory

> CHERI-aware allocators must ignore, or trap on, calls to `free()` if the

Reviewer 2: ```suggestion
CHERI-aware allocators must ignore, or diagnose, calls to `free()` if the
```
Couldn't we also have other behaviours such as logging errors?

Author: I wanted to specifically call out ignoring or failing stop -- but not, for example, corrupting their internal metadata. But I agree that we should allow diagnostics. I'll try rephrasing this a bit.

passed capability:

 * Is invalid (i.e, with its tag bit unset); or
 * Is sealed; or
 * Has bounds that have been changed from those returned by the allocator
   when the allocation was first made; or
 * Has permissions that disallow any of data load, data store, capability
   load, or capability store

The allocator may:

 * Implement fail-stop semantics if the call fails for one or more of the
   above reasons.

### Reallocating memory

The allocator must:

 * Fail a call to `realloc()` if the passed capability violates any of the
   properties checked for in `free()`.
 * Return a capability with the same properties as those defined for
   `malloc()`.

The allocator must not:

>  * Return a new pointer from `realloc()` that has an identical address to the

Reviewer 2: I think this needs some rationale: I imagine the argument for this is that existing software does not handle this case correctly? Is there any reason otherwise that we should ban in-place growing?

Author: To my mind, this is about the fact that pointer equality as defined in CHERI C/C++ compares only the address, and not metadata, and so a equality test by the caller to see if the newly returned pointer should be used won't work, and they may will continue to use the old pointer value. Maybe rather than disallowing this, we should make this a 'should' with more rationale...?

Author: I have added notes on this in new rationale text.
